r2/r2/lib/providers/search/solrsearch.py

```diff
@@ -0,0 +1,742 @@
+
+
+        hits = response['response']['numFound']
+        docs = [doc['id'] for doc in response['response']['docs']]
+        facets = {}
+        facet_fields = response['facet_counts'].get('facet_fields', {})
+        for field in facet_fields:
+            facets[field] = []
+            while facet_fields[field]:
+                value = facet_fields[field].pop(0)
+                count = facet_fields[field].pop(0)
+                facets[field].append(dict(value=value, count=count)) 
+            
+
+        results = Results(docs, hits, facets)
+        return results
+
+
+class LinkSearchQuery(SolrSearchQuery):
+    search_api = g.solr_search_host
+    sorts = {'relevance': 'score',
+             'hot': 'hot2',
+             'top': 'top',
+             'new': 'timestamp',
+             'comments': 'num_comments',
+             }
+    sorts_menu_mapping = {'relevance': 1,
+                          'hot': 2,
+                          'new': 3,
+                          'top': 4,
+                          'comments': 5,
+                          }
+    recents = {
+        'hour': timedelta(hours=1),
+        'day': timedelta(days=1),
+        'week': timedelta(days=7),
+        'month': timedelta(days=31),
+        'year': timedelta(days=366),
+        'all': None,
+        None: None,
+    }
+    known_syntaxes = ("solr", "plain")
+    default_syntax = "solr"
+
+    def customize_query(self, bq):
+        queries = [bq]
+        subreddit_query = self._get_sr_restriction(self.sr)
+        if subreddit_query:
+            queries.append(subreddit_query)
+        if self.recent:
+            recent_query = self._restrict_recent(self.recent)
+            queries.append(recent_query)
+        return self.create_boolean_query(queries)
+
+    @classmethod
+    def create_boolean_query(cls, queries):
+        '''Return an AND clause combining all queries'''
+        if len(queries) > 1:
+            bq = ' AND  '.join(['(%s)' %q for q in queries]) 
+        else:
+            bq = queries[0]
+        return bq
+
+    @staticmethod
+    def _restrict_recent(recent):
+        now = datetime.now(g.tz)
+        since = epoch_seconds(now - recent)
+        return 'timestamp:%i..' % since
+
+    @staticmethod
+    def _get_sr_restriction(sr):
+        '''Return a solr-appropriate query string that restricts
+        results to only contain results from self.sr
+        
+        '''
+        bq = []
+        if (not sr) or sr == All or isinstance(sr, DefaultSR):
+            return None
+        elif isinstance(sr, MultiReddit):
+            for sr_id in sr.sr_ids:
+                bq.append("sr_id:%s" % sr_id)
+        elif isinstance(sr, DomainSR):
+            bq = ["site:'%s'" % sr.domain]
+        elif sr == Friends:
+            if not c.user_is_loggedin or not c.user.friends:
+                return None
+            # The query limit is roughly 8k bytes. Limit to 200 friends to
```

Comment:
Do Solr instances have a size limit for queries? Is it the same as Cloudsearch's (8k)?